test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/ServerSentEventsParserTests.cs

```diff
@@ -0,0 +1,234 @@
+        [InlineData("data: T\r\ndata: Hello\n, World\r\n\r\n", "A '\\n' character can only be used as a line ending")]
+        [InlineData("data: data: \r\n", "Unknown message type: 'd'")]
+        public void ParseSSEMessageFailureCases(string encodedMessage, string expectedExceptionMessage)
+        {
+            var buffer = Encoding.UTF8.GetBytes(encodedMessage);
+            var readableBuffer = ReadableBuffer.Create(buffer);
+            var parser = new ServerSentEventsMessageParser();
+            var consumed = new ReadCursor();
+            var examined = new ReadCursor();
+
+            var ex = Assert.Throws<FormatException>(() => { parser.ParseMessage(readableBuffer, out consumed, out examined, out Message message); });
+            Assert.Equal(expectedExceptionMessage, ex.Message);
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("data:")]
+        [InlineData("data: \r")]
+        [InlineData("data: T\r\nda")]
+        [InlineData("data: T\r\ndata:")]
+        [InlineData("data: T\r\ndata: Hello, World")]
+        [InlineData("data: T\r\ndata: Hello, World\r")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\r")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\r\\")]
+        [InlineData("data: T\r\ndata: Major\r\ndata:  Key\rndata:  Alert\r\n\r\\")]
+        [InlineData("data: T\r\ndata: Major\r\ndata:  Key\r\ndata:  Alert\r\n\r\\")]
+        public void ParseSSEMessageIncompleteParseResult(string encodedMessage)
+        {
+            var buffer = Encoding.UTF8.GetBytes(encodedMessage);
+            var readableBuffer = ReadableBuffer.Create(buffer);
+            var parser = new ServerSentEventsMessageParser();
+            var consumed = new ReadCursor();
+            var examined = new ReadCursor();
+
+            var parseResult = parser.ParseMessage(readableBuffer, out consumed, out examined, out Message message);
+
+            Assert.Equal(ServerSentEventsMessageParser.ParseResult.Incomplete, parseResult);
+        }
+
+        [Theory]
+        [InlineData("d", "ata: T\r\ndata: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T", "\r\ndata: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r", "\ndata: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r\n", "data: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r\nd", "ata: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r\ndata: ", "Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r\ndata: Hello, World", "\r\n\r\n", "Hello, World")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n", "\r\n", "Hello, World")]
+        [InlineData("data: T", "\r\ndata: Hello, World\r\n\r\n", "Hello, World")]
+        [InlineData("data: ", "T\r\ndata: Hello, World\r\n\r\n", "Hello, World")]
```

Comment:
As mentioned above, you'll encounter an issue with something like `"data:T\r\n", "data: Foo\r\n\r\n"` because you don't store the message type outside the parse function. It's masked right now because you always set `messageType` back to `Text`, so a test like `"data: B\r\n", "data: [base64]\r\n\r\n"` will show the issue right away because you'll get a successful parse but it will be a Text frame instead of a Binary frame.